### ISSUE TITLE
Mathjax on linkedom...

### ISSUE THREAD

**Issue by swamidass** [mathjax/MathJax#2833]
Has anyone figured out how to use MathJax on a linkedom object?

There are some tools for applying it to a jsdom (e.g. https://github.com/pkra/mathjax-node-page/), but I'd prefer to use linkedom. https://github.com/WebReflection/linkedom

Linkedom is a much faster alternative to jsdom, https://webreflection.medium.com/linkedom-a-jsdom-alternative-53dd8f699311 , with nearly the same functionality.

I'm hoping to have a way to use Mathjax in a serverside node process on a linkedom object, modifying the dom in place to render the math. Less useful, but also valuable, would be a commandline utility like mjpage to do the same, but using a linkedom object internally for speed.

**Comment by dpvc** [comment]
MathJax v3 is set up to allow you to use alternative DOM implementations, and so it is not hard to get it to work with linkedom.  MathJax uses a "DOM adaptor" to concert the actions it needs into the corresponding actions on whatever DOM is being used, so one only needs to supply the DOM adaptor.  Since linkedom implements much of the browser DOM API, you can piggyback off the current HTMLadaptor to do most of the work, just like the jsdom adaptor does.

In this case, there are a couple of additional issues to take into account (like if you pass an empty string to linkedom's `parseHTML()`, you don't get a fully instantiated DOM, and it doesn't implement the `HTMLCollection` class), but these can be handled by overriding the appropriate methods in the adaptor.

I have pushed a version that seems to work (but I haven't tested it much) in the `issue2833` branch of mathjax/MathJax-src.  You can check out that branch and compile the adaptor using `npm run -s compile`.  Then you can use the new adaptor in place of the jsdom adaptor in the [jsdom examples](https://github.com/mathjax/MathJax-demos-node/tree/master/jsdom) in the [MathJax node demos](https://github.com/mathjax/MathJax-demos-node) repository.

To do so, replace

    const {jsdomAdaptor} = require('mathjax-full/js/adaptors/jsdomAdaptor.js');

with

    const {linkedomAdaptor} = require('mathjax-full/js/adaptors/linkedomAdaptor.js');

and

    const {JSDOM} = require('jsdom');

with

    const {parseHTML} = require('linkedom');

and finally

    const adaptor = jsdomAdaptor(JSDOM);

with

    const adaptor = linkedomAdaptor(parseHTML);

and that should allow any fo the jsdom examples to run, including the ones that process a whole page at once.

**Comment by swamidass** [comment]
Thanks for this. I'll look into it and see. When do you think that will be merged into main branch of MathJax?

**Comment by dpvc** [comment]
I expect it to be in the next release, but I don't think that will be for several months yet.  There is a lot still to go into it.

**Comment by dpvc** [comment]
The [jsdom examples](https://github.com/mathjax/MathJax-demos-node/tree/master/jsdom) that I pointed you to earlier show how to work both on a single expression and on a complete DOM.  The latter are the examples whose names end in `-page`.

The only difference to make it work on an existing DOM (rather than a file, like in the examples), is to pass the linkedom Document as the first argument to `mathjax.document()` rather than the HTML string.  I gave you the other needed changes in my response above.

Give that a try.
